packages/plugin-ext/src/main/browser/plugin-manager-client.ts

```diff
@@ -157,28 +157,32 @@ export class HostedPluginManagerClient {
      * Creates directory choose dialog and set selected folder into pluginLocation field.
      */
     async selectPluginPath(): Promise<void> {
-        const root = await this.workspaceService.root || await this.fileSystem.getCurrentUserHome();
-        if (!root) {
-            throw new Error('Unable to find the root');
+        let rootStat = await this.workspaceService.root;
+
+        if (!rootStat) {
+            rootStat = await this.fileSystem.getCurrentUserHome();
         }
-        const rootUri = new URI(root.uri);
-        const rootStat = await this.fileSystem.getFileStat(rootUri.toString());
+
         if (!rootStat) {
-            throw new Error('Unable to find the rootStat');
+            throw new Error('Unable to find the root');
         }
 
-        const name = this.labelProvider.getName(rootUri);
-        const label = await this.labelProvider.getIcon(root);
+        const name = this.labelProvider.getName(rootStat);
+        const label = await this.labelProvider.getIcon(rootStat);
         const rootNode = DirNode.createRoot(rootStat, name, label);
         const dialog = this.fileDialogFactory({ title: HostedPluginCommands.SELECT_PATH.label! });
```

Comment:
Can we somehow trick ts to return a result of different types depending on incoming dialog parameters? i.e. if `multiSelect` is not true, then it should be a single file select dialog that clients don't need to check for other cases. Type guards below are redundant, woud be nice to remove unnecessary code.

Reply:
Reworked a little bit. Could you please give a look again?